library/asn1parse.c

```diff
@@ -163,7 +163,7 @@ int mbedtls_asn1_get_int( unsigned char **p,
     *val = 0;
     while( len-- > 0 )
     {
-        *val = ( *val << 8 ) | **p;
+        *val = ( ((unsigned int) *val) << 8 ) | **p;
```

Comment:
That results in the wrong value when the value fits in an `unsigned int` but not in an `int`, doesn't it?

Reply:
The undefined behaviour is one thing, but the entire function looks ... tricky.

If I was to reverse engineer the ASN1 INTEGER spec from this code I would have to deduce that:
1. an empty byte string is not a valid ASN1 integer.
2. the value zero can be represented as any number (>=1) of `00` bytes.
3. the bytes `00 00 00 00 00 00 00 00 00 00 12 34 56 78` are a valid ASN1 integer, with the value `0x12345678`
4. the byte `87` is not a valid integer in ASN1, presumably 0x87 has to rendered as `00 87`?
5. the bytes `87 65 43 21` are not a valid integer in ASN1
6. the bytes `00 87 65 43 21` are a valid ASN1 integer in ASN1, with 'value' 0x87654321. Which is negative.
7. the bytes `12 34 56 78 00` are not a valid integer in ASN1.

The strangest artefact of the implementation is the `(**p & 0x80) != 0` test which rejects values that have msb set (e.g. 4 above). But it doesn't reject values >=2<sup>31</sup> which in our systems are cast to negative signed integer values (see 6 above).

Reply:
> now I've read teh ASN.1 DER encoding rules

Commiserations.

> is too relaxed about leading zeros

We don't necessarily reject non-BER encodings. It's a matter of interoperability. I'd like to tighten things up, because [being liberal in what you accept is not such a good idea](https://tools.ietf.org/id/draft-thomson-postel-was-wrong-03.html), but we can only do that after a careful review to ensure we don't [break somebody's workflow](https://xkcd.com/1172/).

> errors on correctly represented negative values

This is cryptography. We don't deal in negative values. The fact that the encoding supports them is an unfortunate historical relic.

> returns a negative result for correctly formatted postive integers

Or rather, since this is a signed int, triggered undefined behavior, which in many environments would manifest itself by returning a negative result. The correct behavior is actually to return an error since the number is out of the representable range in this case.

Reply:
> If I was to reverse engineer the ASN1 INTEGER spec

Note that we don't exactly implement the ASN.1 specification or even a subset of it. We implement something that is close to a subset of it, but that works in the real world which is a vicious circle of people not quite implementing the specification correctly and the rest having to follow through for interoperability and backward compatibility.